src/main/java/com/direwolf20/buildinggadgets/common/items/gadgets/GadgetDestruction.java

```diff
@@ -383,33 +387,41 @@ public void clearArea(World world, BlockPos pos, EnumFacing side, EntityPlayer p
         NBTTagCompound tagCompound = new NBTTagCompound();
-        List<Integer> posIntArrayList = new ArrayList<Integer>();
-        List<Integer> stateIntArrayList = new ArrayList<Integer>();
-        List<Integer> pastePosArrayList = new ArrayList<Integer>();
-        List<Integer> pasteStateArrayList = new ArrayList<Integer>();
+        IntList posIntArrayList = new IntArrayList();
+        List<IBlockState> intStateArrayList = new ArrayList<>();
+        IntList stateIntArrayList = new IntArrayList();
+        IntList pastePosArrayList = new IntArrayList();
+        IntList pasteStateArrayList = new IntArrayList();
         BlockMapIntState blockMapIntState = new BlockMapIntState();
         String UUID = getUUID(stack);
 
         for (Map.Entry<BlockPos, IBlockState> entry : posStateMap.entrySet()) {
             posIntArrayList.add(GadgetUtils.relPosToInt(startBlock, entry.getKey()));
             blockMapIntState.addToMap(entry.getValue());
             stateIntArrayList.add((int) blockMapIntState.findSlot(entry.getValue()));
+            intStateArrayList.add(entry.getValue());
             if (pasteStateMap.containsKey(entry.getKey())) {
                 pastePosArrayList.add(GadgetUtils.relPosToInt(startBlock, entry.getKey()));
                 IBlockState pasteBlockState = pasteStateMap.get(entry.getKey());
                 blockMapIntState.addToMap(pasteBlockState);
                 pasteStateArrayList.add((int) blockMapIntState.findSlot(pasteBlockState));
             }
         }
-        tagCompound.setTag(NBTKeys.MAP_STATE_INT, blockMapIntState.putIntStateMapIntoNBT());
-        int[] posIntArray = posIntArrayList.stream().mapToInt(i -> i).toArray();
-        int[] stateIntArray = stateIntArrayList.stream().mapToInt(i -> i).toArray();
-        int[] posPasteArray = pastePosArrayList.stream().mapToInt(i -> i).toArray();
-        int[] statePasteArray = pasteStateArrayList.stream().mapToInt(i -> i).toArray();
-        tagCompound.setIntArray(NBTKeys.MAP_POS_INT, posIntArray);
-        tagCompound.setIntArray(NBTKeys.MAP_STATE_INT, stateIntArray);
-        tagCompound.setIntArray(NBTKeys.MAP_POS_PASTE, posPasteArray);
-        tagCompound.setIntArray(NBTKeys.MAP_STATE_PASTE, statePasteArray);
+        tagCompound.setTag(NBTKeys.MAP_INDEX2STATE_ID, blockMapIntState.putIntStateMapIntoNBT());
+        tagCompound.setIntArray(NBTKeys.MAP_INDEX2POS, posIntArrayList.toIntArray());
+        tagCompound.setIntArray(NBTKeys.MAP_INDEX2STATE_ID, stateIntArrayList.toIntArray());
+        {
+            NBTTagList intStateNBT = new NBTTagList();
+            for (int i = 0; i < intStateArrayList.size(); i++) {
+                NBTTagCompound entry = new NBTTagCompound();
+                entry.setInt(NBTKeys.MAP_SLOT, i);
+                entry.setTag(NBTKeys.MAP_STATE, NBTUtil.writeBlockState(intStateArrayList.get(i)));
+                intStateNBT.add(entry);
+            }
+            tagCompound.setTag(NBTKeys.MAP_PALETTE, intStateNBT);
```

Comment:
There should already be a utility Method for that...
If not I'll write it in a few min

Reply:
Sorry but I can't to find it.

Reply:
Ok